### PR TITLE
encode: escape non-ASCII characters while marshalling

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Ref. https://github.com/nspcc-dev/neo-go/pull/2174.

### CODE REVIEW

File: decode.go [hunk 1]
```diff
@@ -1252,13 +1252,19 @@ func unquoteBytes(s []byte) (t []byte, ok bool) {
 					if dec := utf16.DecodeRune(rr, rr1); dec != unicode.ReplacementChar {
 						// A valid pair; consume.
 						r += 6
-						w += utf8.EncodeRune(b[w:], dec)
+						// Do not replace with unescaped rune, need to follow reference behaviour,
+						// see https://github.com/nspcc-dev/neo-go/pull/2174.
```
Comment (roman-khimov) [inline decode.go:1256]: It's not correct. We need to unescape strings to get correct ones and C# sure does the same thing.

File: encode.go [hunk 2]
```diff
@@ -913,7 +914,13 @@ func (e *encodeState) string(s string, escapeHTML bool) int {
 				e.WriteString(s[start:i])
 			}
 			switch b {
-			case '\\', '"':
+			case '\\':
+				if i+1 < l && s[i+1] == 'u' {
```
Comment (roman-khimov) [inline encode.go:918]: This is wrong, we need to escape things C# normally escapes, but this just breaks serialization (I can legitimately have "\uxxxx" string that I want to be correctly JSONized and passed to other party).

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module github.com/nspcc-dev/go-ordered-json
+
+go 1.17
```
Comment (roman-khimov) [inline go.mod:3]: 1.15?

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+
+require (
+	github.com/virtuald/go-ordered-json v0.0.0-20170621173500-b18e6e673d74
```
Comment (roman-khimov) [inline go.mod:6]: Add `rewrite` to get it from `nspcc-dev/go-ordered-json`?

Comment (AnnaShaleva) [reply go.mod:6]: Isn't `rewrite` should be added to neo-go?

Comment (AnnaShaleva) [reply go.mod:6]: Fixed.

**Comment by AnnaShaleva** [comment]
@roman-khimov, I added README update in a separate PR not to mix the things, see #2.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**